src/Specular/FRP/Base.purs

```diff
@@ -234,41 +232,53 @@ current = Behavior
 
 -- | An Event that fires with the new value every time the Dynamic changes.
 changed :: forall a. Dynamic a -> Event a
-changed (Dynamic node) = Event node
+changed (DynPure _) = never
+changed (DynNode node) = Event node
 
 -- | An Event that fires every time the Dynamic changes.
 changed_ :: forall a. Dynamic a -> Event Unit
 changed_ = changed <<< void
 
 instance functorDynamic :: Functor Dynamic where
-  map f (Dynamic node) = Dynamic $ unsafePerformEffect do
+  map f (DynPure x) = DynPure (f x)
+  map f (DynNode node) = DynNode $ unsafePerformEffect do
     n <- runEffectFn2 I.map f node
     runEffectFn2 Node.annotate n ("map " <> Node.name node)
     pure n
 
 instance applyDynamic :: Apply Dynamic where
-  apply (Dynamic f) (Dynamic x) = Dynamic $ unsafePerformEffect do
+  apply (DynPure f) dx = f <$> dx
+  apply df (DynPure x) = (_ $ x) <$> df
+  apply (DynNode f) (DynNode x) = DynNode $ unsafePerformEffect do
     n <- runEffectFn3 I.map2 (mkFn2 ($)) f x
     runEffectFn2 Node.annotate n ("apply (" <> Node.name f <> ") (" <> Node.name x <> ")")
     pure n
 
 map2 :: forall a b c. (a -> b -> c) -> Dynamic a -> Dynamic b -> Dynamic c
-map2 f (Dynamic x) (Dynamic y) = Dynamic $ unsafePerformEffect do
+map2 f (DynPure x) dy = f x <$> dy
+map2 f dx (DynPure y) = (\x -> f x y) <$> dx
+map2 f (DynNode x) (DynNode y) = DynNode $ unsafePerformEffect do
   n <- runEffectFn3 I.map2 (mkFn2 f) x y
   runEffectFn2 Node.annotate n ("map2 (" <> Node.name x <> ") (" <> Node.name y <> ")")
   pure n
 
 mapN :: forall a b. (Array a -> b) -> Array (Dynamic a) -> Dynamic b
-mapN f inputs = Dynamic $ unsafePerformEffect do
-  n <- runEffectFn2 I.mapN f (coerce inputs)
+mapN f inputs = DynNode $ unsafePerformEffect do
+  -- TODO: optimize (eliminate pure nodes)
+  let inputs' = dynToNode <$> inputs
+  n <- runEffectFn2 I.mapN f inputs'
   runEffectFn2 Node.annotate n "mapN"
   pure n
 
-instance applicativeDynamic :: Applicative Dynamic where
-  pure x = Dynamic $ unsafePerformEffect do
-    n <- runEffectFn1 I.constant x
-    runEffectFn2 Node.annotate n "pure"
-    pure n
+instance Applicative Dynamic where
+  pure = DynPure
+
+dynToNode :: forall a. Dynamic a -> Node a
+dynToNode (DynPure x) = unsafePerformEffect do
```

Comment:
It's interesting: Dynamic Apply instance doesn't need this, Dynamic Bind instance does. And `mapAsync` does. It seems this should be avoidable so we have even less nodes. Or it isn't?